_locate_cmd: use shlex.join to retain word boundaries in nrpe check commands

jetpackdanger opened the pull request:
Addresses https://github.com/juju/charm-helpers/issues/408 and https://bugs.launchpad.net/charm-helpers/+bug/1706265

jetpackdanger commented:
Ugh, looks like shlex.join is too new.

ajkavanagh commented on charmhelpers/contrib/charmsupport/nrpe.py:
```diff
@@ -122,6 +122,34 @@
+    def shlex_join(split_command):
+        """Return a shell-escaped string from *split_command*."""
+        return ' '.join(shlex_quote(arg) for arg in split_command)
+
```
It feels like there should be a unit test for this to verify that it is doing what the Py3.8 version does?

ajkavanagh commented on charmhelpers/contrib/charmsupport/nrpe.py:
```diff
@@ -122,6 +122,34 @@
 
 
+# If shlex.join is not available (new as of Python 3.8) then let's emulate it.
+if 'join' in dir(shlex):
```
Maybe use:

```python
try:
    shlex_join = shlex.join
except AttributeError:  # it might be something else?
   # rest of code, etc.
```

ajkavanagh commented on charmhelpers/contrib/charmsupport/nrpe.py:
```diff
@@ -122,6 +122,34 @@
+        # Python 3 defaults to Unicode
+        _find_unsafe = re.compile(r'[^\w@%+=:,./-]', re.ASCII).search
+    else:
+        # Python 2 defaults to ASCII
```
We've got `six` available so that a test of `if six.PY3: ...` can be done.

jetpackdanger commented:
Thank you for all the excellent suggestions!

However, I've just realized there's a fatal flaw with this approach, which is that all of the existing `add_check` calls will  get an extra level of escaping that will probably break them. For example:

    >>> command = '/usr/lib/nagios/plugins/check_http -k "Accept:\\ application/json" -I 127.0.0.1 -p 9000 -u /api -s cluster_id'
    >>> print(' '.join(shlex.split(command)))  # currently
    /usr/lib/nagios/plugins/check_http -k Accept:\ application/json -I 127.0.0.1 -p 9000 -u /api -s cluster_id
    >>> print(shlex.join(shlex.split(command)))  # with my change
    /usr/lib/nagios/plugins/check_http -k 'Accept:\ application/json' -I 127.0.0.1 -p 9000 -u /api -s cluster_id
    >>> _

I haven't been able to think of a clever idea to detect such cases yet, so I'll  close this PR and think about it some more.
